### PR TITLE
[FIX] account: Some fixes since account-pocalypse

### PR DESCRIPTION
**Pull request by smetl**
1) Improve readability / performance of moves checks

The queries checking the move name & the duplicate vendor reference was complex
and unefficient for no reason.

2) Recompute amount_currency when reversing invoice

When reversing an invoice, the amount_currency wasn't recomputed even when dealing with
a refund date having another currency rate.

3) The refund must not keep the payment ref of the reverted invoice.

The payment ref is stored inside the receivable/payable line's name. However, when
duplicating the invoice to create a refund, this must be emptied until a new
payment ref will be set.

4) Reintroduce user_id field on account.move

The field user_id has been renamed to invoice_user_id. However, some mail templates
need to access the generic user_id field for multiple models.

5) Fix 'misc' Journal Entry totals

Misc journal entry must have an amount total always positive and not signed.
Furthermore, we don't know how to compute the amount_untaxed/amount_tax so don't
compute that with such entries.

6) Fix cancelling an exchange difference entry / tax cash basis entry.

When cancelling an exchange difference entry, the reconciliation is undo and then,
the exchange difference entry is reversed. Same for tax cash basis entries.

issue: #35841

7) Fix creation of cash basis entries for reversed journal entries.

When reconciling entries with one being the reverse of the another one, cash basis entries
for taxes must not be created. However, it must be the case when reconciling the reverse entry
(like refunds) with something.

issue: #35276

8) Fix matching rules ignoring the structured communication.

Since account-pocalypse, the structured communication is landing on the name of the
journal items. Then, the matching rules must consider it.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_move.py [hunk 1]
```diff
@@ -1238,25 +1238,28 @@ def _validate_move_modification(self):
 
     @api.constrains('name', 'journal_id')
```
Comment (jco-odoo) [inline addons/account/models/account_move.py]: state should be added

File: addons/account/models/account_move.py [hunk 2]
```diff
@@ -1238,25 +1238,28 @@ def _validate_move_modification(self):
...
 
+        values_table = ','.join('(%s, %s, %s, %s, %s)' for move in moves)
         self._cr.execute('''
-            WITH current_moves (id, name, state, company_id, journal_id, type) as ( values {} )
-            SELECT move2.id
-            FROM (    SELECT id, name, state, company_id, journal_id, type from account_move
-            UNION ALL SELECT id, name, state, company_id, journal_id, type from current_moves
-            ) move
+            SELECT move.id
+            FROM (VALUES ''' + values_table + ''') AS move(id, name, state, type, journal_id)
```
Comment (jco-odoo) [inline addons/account/models/account_move.py]: no way to put this in the where-params?

File: addons/account/models/account_move.py [hunk 3]
```diff
@@ -1267,23 +1270,22 @@ def _check_duplicate_supplier_reference(self):
...
+
+        values_table = ','.join('(%s, %s, %s, CAST(%s AS DATE), %s, %s)' for move in moves)
         self._cr.execute('''
-            WITH current_moves (id,  ref, company_id, commercial_partner_id, type, invoice_date) as ( values {} )
-            SELECT move2.id
-            FROM (    SELECT id, ref, company_id, commercial_partner_id, type, invoice_date from account_move
-            UNION ALL SELECT id, ref, company_id, commercial_partner_id, type, invoice_date from current_moves
-            ) move
+            SELECT move.id
+            FROM (VALUES ''' + values_table + ''') AS move(id, ref, type, invoice_date, company_id, commercial_partner_id)
```
Comment (jco-odoo) [inline addons/account/models/account_move.py]: idem

File: addons/account/models/account_move.py [hunk 4]
```diff
@@ -3385,6 +3394,10 @@ def _copy_data_extend_business_fields(self, values):
         '''
         self.ensure_one()
 
+        # Don't copy the name of a payment term line.
+        if self.move_id.is_invoice() and self.account_id.user_type_id.type in ('receivable', 'payable'):
+            values['name'] = ''
+
```
Comment (jco-odoo) [inline addons/account/models/account_move.py]: Hm, it says include_business_fields and we don't copy them.  Should not it go in the method below?

File: addons/account/models/account_move.py [hunk 5]
```diff
@@ -3535,6 +3535,11 @@ def _query_get(self, domain=None):
         if not isinstance(domain, (list, tuple)):
             domain = safe_eval(domain)
 
+        domain += [
+            ('display_type', 'not in', ('line_section', 'line_note')),
+            ('move_id.state', '!=', 'cancel'),
+        ]
+
```
Comment (jco-odoo) [inline addons/account/models/account_move.py]: Weird, it is already there below (and even then I think it is weird to override this so extensively, seems like a rather low-level method)

File: addons/account/models/account_move.py [hunk 6]
```diff
@@ -1344,7 +1344,9 @@ def _check_fiscalyear_lock_date(self):
 
     @api.multi
     def _check_tax_lock_date(self):
```
Comment (jco-odoo) [inline addons/account/models/account_move.py]: Is it called from a constraint somewhere?

File: addons/account/models/account_move.py [hunk 7]
```diff
@@ -1354,7 +1356,7 @@ def _check_tax_lock_date(self):
             JOIN res_company company ON company.id = journal.company_id
             WHERE move.id IN %s
             AND move.date < company.tax_lock_date
-        ''', [tuple(self.ids)])
+        ''', [tuple(moves.ids)])
```
Comment (jco-odoo) [inline addons/account/models/account_move.py]: Is the term clear?  Because this would mean that someone could still add something the the sales revenue without taxes while the sales revenue needs to be reported e.g.  You seem to give a lot more freedom here.

File: addons/account/models/account_move.py (hunk 2 above)
Comment (smetl) [reply addons/account/models/account_move.py]: params are already in where_params....

Comment (jco-odoo) [reply addons/account/models/account_move.py]: No way to put values_table in the 2nd arg of the execute method instead of concatenating strings?

File: addons/account/models/account_move.py (hunk 6 above)
Comment (smetl) [reply addons/account/models/account_move.py]: in post

**Comment by smetl** [comment]
@jco-odoo A little r+ maybe? :)

**Comment by jco-odoo** [comment]
@robodoo delegate+

File: addons/account/models/account_move.py [hunk 8]
```diff
@@ -1267,23 +1270,22 @@ def _check_duplicate_supplier_reference(self):
...
+            where_params += [move.id, move.ref, move.type, move.invoice_date or None, move.company_id.id, move.commercial_partner_id.id]
+
+        values_table = ','.join('(%s, %s, %s, CAST(%s AS DATE), %s, %s)' for move in moves)
         self._cr.execute('''
-            WITH current_moves (id,  ref, company_id, commercial_partner_id, type, invoice_date) as ( values {} )
-            SELECT move2.id
-            FROM (    SELECT id, ref, company_id, commercial_partner_id, type, invoice_date from account_move
-            UNION ALL SELECT id, ref, company_id, commercial_partner_id, type, invoice_date from current_moves
-            ) move
+            SELECT move.id
```
Comment (oco-odoo) [inline addons/account/models/account_move.py]: move2.id would be better I guess

File: addons/account/models/account_move.py [hunk 9]
```diff
@@ -1236,27 +1236,30 @@ def _validate_move_modification(self):
...
             return
 
+        values_table = ','.join('(%s, %s, %s, %s, %s)' for move in moves)
         self._cr.execute('''
-            WITH current_moves (id, name, state, company_id, journal_id, type) as ( values {} )
-            SELECT move2.id
-            FROM (    SELECT id, name, state, company_id, journal_id, type from account_move
-            UNION ALL SELECT id, name, state, company_id, journal_id, type from current_moves
-            ) move
+            SELECT move.id
```
Comment (oco-odoo) [inline addons/account/models/account_move.py]: move2.id

File: addons/account/models/account_move.py [hunk 10]
```diff
@@ -3388,8 +3397,11 @@ def _copy_data_extend_business_fields(self, values):
     @api.multi
     def copy_data(self, default=None):
         res = super(AccountMoveLine, self).copy_data(default=default)
-        if self._context.get('include_business_fields'):
-            for line, values in zip(self, res):
+        for line, values in zip(self, res):
```
Comment (oco-odoo) [inline addons/account/models/account_move.py]: this fix is not related to the title of the commit it belongs to; make an additionnal commit or squash everything

File: addons/account/models/account_move.py [hunk 11]
```diff
@@ -1344,7 +1344,9 @@ def _check_fiscalyear_lock_date(self):
 
     @api.multi
     def _check_tax_lock_date(self):
-        if not self:
+        moves = self.filtered(lambda move: any(line.tax_repartition_line_id or line.tax_ids for line in move.line_ids))
```
Comment (oco-odoo) [inline addons/account/models/account_move.py]: What about tax adjustments?
=> We should also check the tags.

Change to check the tax lock date if the aml hax tax_ids or tax_repartition_line_id, or if some of its tag_ids have applicability=='taxes'

Comment (smetl) [reply addons/account/models/account_move.py]: to be fixed on saas-12.3

**Comment by smetl** [comment]
@robodoo r+
